### ISSUE TITLE
Exclusion and coordinated decision points

### ISSUE THREAD

**Issue by SritterCL** [CarnegieLearningWeb/UpGrade#245]
The experiment designer should have the ability to describe how exclusion works for experiments with coordinated decision points. Suppose there is an experiment with 3 coordinated decision points. Some experiments would only want to collect data from students who visited all three decision points (for example, the first could be a pretest, the second an educational intervention and the third a posttest - if students don't do all three, their data isn't going to be useful). So, in this case, UpGrade should exclude students who visited point 1 before the experiment started.

Alternatively, we could have an experiment where we're testing the same thing at three different decision points (e.g. we're testing motivational messages in three different workspaces). In that case, it's OK for a student to experience only 1 of the three decision points. So, in this kind of experiment we don't want to exclude a student who visited point 1 before the experiment started.

Basically, we need to make this behavior a parameter.

I think the current behavior is to exclude students who reached at least 1 of the decision points (Vivek agreed). But we need to test this.

**Comment by SritterCL** [comment]
(and, I suppose, for experiments that only have one decision point, we could still allow the experimenter to specify that users who have previously reached that point are *not* excluded)

**Comment by ppratikcr7** [comment]
@SritterCL It is working as expected. If a user does mark a single decision point then all 3 coordinated decision points get excluded on getAllExperimentConditions.

**Comment by amurphy-cl** [comment]
If this is resolved, can we close the issue?

**Comment by ppratikcr7** [comment]
As per the ticket description the current behaviour is verified but it also states that we should have a parameter to change this behaviour as an enhancement change.

**Comment by nirmalpatel** [comment]
@SritterCL We were discussing this issue, and there seem to be two user stories in this issue:

1) Which all decision points should a user hit to be considered enrolled in the experiment?
2) Should users be excluded if they have visited one or more DPs within the experiment?

We are not sure how the second one relates to the Experiment consistency rule.

Can you share how you are thinking about the exclusion behavior here?

**Comment by amurphy-cl** [comment]
Hey Nirmal, Steve and I talked about this, and the behavior should be:

1. If all decision points are required, reaching _any_ decision point enrolls the user in the experiment 
2. The “required” (or whatever we call it, probably need a better name) field is what is used to define whether users are excluded from the experiment if they’ve visited any such decision point before the experiment starts running. I think if the consistency rule is set to Experiment, that will override decision point settings. 

Here's a link to a google sheet that (hopefully) lays out the logic more clearly 
https://docs.google.com/spreadsheets/d/1BHxVBm4LWD7k8uSltIKDp5rhVlJHs99j6XWMm2QUF2k/edit#gid=0

**Comment by SritterCL** [comment]
April and I re-worked that spreadsheet to (hopefully) make clearer what's going on, so take a look again: https://docs.google.com/spreadsheets/d/1BHxVBm4LWD7k8uSltIKDp5rhVlJHs99j6XWMm2QUF2k/edit#gid=609987306

**Comment by SritterCL** [comment]
Also, if we have "required" checkboxes for decision points, even in single-decision-point experiments, we don't think we need Experiment consistency.

**Comment by nirmalpatel** [comment]
Thanks for the examples!

I guess when no Required checkboxes are selected, then it becomes Experiment consistency, and when all are checked, it becomes Individual consistency... let me know if this is not correct...

**Comment by amurphy-cl** [comment]
I think so, as long as we all understand that condition assignment for the Consistency Rule is with respect to the _entire experiment_, not each decision point.
